pkg/proxy/iptables/proxier.go

```diff
@@ -1093,6 +1093,33 @@ func (proxier *Proxier) syncProxyRules() {
 				proxier.natChains.Write(utiliptables.MakeChainLine(svcChain))
 			}
 			activeNATChains[svcChain] = true
+
+			if proxier.masqueradeAll {
+				proxier.natRules.Write(
+					"-A", string(svcChain),
+					"-j", string(KubeMarkMasqChain))
+			} else {
+				// All NodePort, ExternalIP, and Load Balancer traffic
+				// (which is to say, all non-ClusterIP traffic) gets
+				// masqueraded.
+				if svcInfo.NodePort() != 0 || len(svcInfo.ExternalIPStrings()) != 0 || len(svcInfo.LoadBalancerIPStrings()) != 0 {
+					proxier.natRules.Write(
+						"-A", string(svcChain),
+						"!", "-d", svcInfo.ClusterIP().String(),
+						"-j", string(KubeMarkMasqChain))
+				}
+
+				// Likewise all traffic to a to ClusterIP from a non-pod
+				// source gets masqueraded (if we know how to distinguish
+				// pods from non-pods).
+				if proxier.localDetector.IsImplemented() {
+					proxier.natRules.Write(
+						"-A", string(svcChain),
+						"-d", svcInfo.ClusterIP().String(),
+						proxier.localDetector.IfNotLocal(),
+						"-j", string(KubeMarkMasqChain))
+				}
```

Comment:
should the not-implemented case here be to masquerade or not?

Reply:
Not. The not-implemented case is when we can't tell whether the traffic came from a pod or not-from-a-pod, and in that case, we assume that all traffic to the clusterIP must have come from a pod. (Because we don't want to force masquerading in the pod-to-clusterIP case just so that we can make the normally-not-even-sued external-to-clusterIP case work.)

(Note that this is exactly the same behavior as before, I've just consolidated the rules.)